docs/RFCS/20171118_role_based_access_control.md

```diff
@@ -0,0 +1,747 @@
+* **Cons**: more complex manipulation of roles, increased cost for very large roles
+
+Dual tables: one for direct role membership information, one for expanded membership information:
+* the direct membership table would include the `ADMIN OPTION` field and would be queried for role manipulation
+* permission checking only requires lookups in the expanded table
+* the expanded table can use an index on the user or even an array type for the list of roles
+* **Pros**: fast lookup of a users's memberships
+* **Cons**: slightly more complex role manipulation
+
+## Unresolved questions
+
+These must be resolved before moving into the final comment period.
+
+### Leases on system tables
+
+We need to check whether leases can be acquired on system tables (specifically, the `role_members` table).
+
+If not, an alternative would be to make the `role_members` table a non-system table, or to use an separate
+empty table to acquire leases on.
+
+### Inheritance of role admin
+
+Given the following:
+* `otherrole ∈ mainrole WITH ADMIN OPTION`
+* `marc ∈ otherrole`
+
+Do we allow inheritance of role admin?
+* if yes: `marc` is allowed to modify the membership of `mainrole`
+* if no: `marc` is not allowed to modify the membership of `mainrole`
+
+We should follow the PostgreSQL logic.
+
+### Adding and removing admin setting
+
+When calling `GRANT ... WITH ADMIN OPTION` or `REVOKE ADMIN OPTION FOR ...`, does this add/remove the member
+or only the admin attribute?
+
+Effectively, this boils down to how these calls should work:
+```
+# Add the user without admin
+GRANT myrole TO marc
+# Does this fail on existing member, or does this make marc an admin?
+GRANT myrole TO marc WITH ADMIN OPTION
+```
+
+And for revoke:
+```
+# Given marc as an admin member of myrole, does this remove marc as a member, or only remove the admin attribute?
+REVOKE myrole FROM marc WITH ADMIN OPTION
+```
+
+We should follow the PostgreSQL logic.
+
+### Listing expanded role memberships
+
+The `SHOW GRANTS ON ROLE` statement allows lookup of role memberships.
+
+We should have a way to show or filter direct vs indirect memberships.
+Some options:
+* include a `DIRECT` field which returns `YES` is we are a direct member
+* add an option to `SHOW GRANTS` to show all/direct/indirect memberships
+
+### Virtual table details
+
+The [Virtual tables](#virtual-tables) section has a few TODOs to clarify PostgreSQL behavior.
+
+### Existence of an `admin` user before migrations
+
+One of the migrations needed for role-based access control is the addition of the `admin` role to the
+`system.users` table.
+
+We should decide what to do if a user with that name exists. Some possibilities:
+* just overwrite it (clear the password hash and set the `isRole` field) and document the upgrade breakage.
+* rename it: this would involve renaming all privileges as well. Surfacing this to the user would be tricky.
+* fail the migration with a message clearly saying the user must be removed/renamed.
```

Comment:
I'd vote for this along with a release note, for what it's worth

Reply:
Except I don't think we have a good way to 1) notify users, and 2) resume the migration after the problem has been fixed.
This would involve the concept of "skippable" and "resumable" migrations. I'm voting for "just overwrite it", mostly because that's by far the safest solution.

Reply:
It's not safe for someone that has an admin user and won't be able to log in as them after the upgrade.

If it changes anything, I interpret "fail the migration" as "crash the server with a `log.Fatal` message explaining what needs to be done".

Reply:
So log fatal and ask them to start with an older binary and remove the user? That's one option albeit a radical one.

Reply:
Or if it's part of a cluster, just connect to one of the already running nodes and do it.